snapcraft/internal/lxd.py

```diff
@@ -183,18 +197,69 @@ def _ensure_container(self):
     def _setup_project(self):
         self._ensure_mount(self._project_folder, self._source)
 
+    def _container_login(self):
+        network = self._get_container_status()['state']['network']['eth0']
+        for address in network['addresses']:
+            if address['family'] == 'inet':
+                return address['address'], '22'
+        raise RuntimeError('No IP found for {}'.format(self._container_name))
+
+    def _background(self, cmd, stdin=None, stdout=None):
+        self._processes += [Popen(cmd, stdin=stdin, stdout=stdout)]
+
     def _ensure_mount(self, destination, source):
         logger.info('Mounting {} into container'.format(source))
-        devices = self._get_container_status()['devices']
-        if destination not in devices:
-            check_call([
-                'lxc', 'config', 'device', 'add', self._container_name,
-                destination, 'disk', 'source={}'.format(source),
-                'path=/{}'.format(destination)])
+        remote, container_name = self._container_name.split(':')
+        if remote != 'local':
+            # Remove project folder in case it was used "locally" before
+            devices = self._get_container_status()['devices']
+            if destination in devices:
+                check_call([
+                    'lxc', 'config', 'device', 'remove', self._container_name,
+                    destination, 'disk', 'source={}'.format(source),
+                    'path=/{}'.format(destination)])
+
+            self._install_packages(['sshfs'])
+            keyfile = 'id_{}'.format(self._container_name)
+            if not os.path.exists(keyfile):
+                check_output(['ssh-keygen', '-o', '-N', '', '-f', keyfile])
+            self._container_run(['mkdir', '-p', '/root/.ssh'])
+            self._container_run(['chmod', '700', '/root/.ssh'])
+            self._container_run(['tee', '-a',
+                                 '/root/.ssh/authorized_keys'],
+                                stdin=open('{}.pub'.format(keyfile), 'r'))
+            self._container_run(['chmod', '600',
+                                 '/root/.ssh/authorized_keys'])
+            self._container_run(['mkdir', '-p', '/{}'.format(destination)])
+            self._container_run(['mkdir', '-p', source])
+            r1, w1 = os.pipe()
+            r2, w2 = os.pipe()
```

Comment:
instead of r1 and w2, why not calling this stdin and stdout?
And when you don't use variables, you use _ instead of the name.
`stdin, _ = os.pipe()`